pkg/operator/defaultscccontroller/defaultscccontroller.go

```diff
@@ -0,0 +1,131 @@
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+const (
+	ControllerName = "default-scc-upgradeable"
+)
+
+type Options struct {
+	Config         *rest.Config
+	OperatorClient v1helpers.OperatorClient
+	Recorder       events.Recorder
+	ResyncPeriod   time.Duration
+}
+
+func NewDefaultSCCController(options *Options) (controller *DefaultSCCController, err error) {
+	client, clientErr := securityv1client.NewForConfig(options.Config)
+	if clientErr != nil {
+		err = fmt.Errorf("[%s] failed to create client for security/v1 - %s", ControllerName, clientErr.Error())
+		return
+	}
+
+	// Create a new SecurityContextConstraints watcher
+	watcher := &cache.ListWatch{
+		ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+			return client.SecurityContextConstraints().List(options)
+		},
+
+		WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+			return client.SecurityContextConstraints().Watch(options)
+		},
+	}
+
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "DefaultSCCController")
+
+	indexer, informer := cache.NewIndexerInformer(watcher, &securityv1.SecurityContextConstraints{}, options.ResyncPeriod, NewResourceHandler(queue), cache.Indexers{})
+	lister := securityv1listers.NewSecurityContextConstraintsLister(indexer)
+	recorder := options.Recorder.WithComponentSuffix(ControllerName)
+
+	controller = &DefaultSCCController{
+		queue:    queue,
+		informer: informer,
+		recorder: recorder,
+		syncer: &Syncer{
+			lister:   lister,
+			recorder: recorder,
+		},
+	}
+
+	return
+}
+
+type DefaultSCCController struct {
+	queue    workqueue.RateLimitingInterface
+	informer cache.Controller
+	recorder events.Recorder
+	syncer   *Syncer
+}
+
+func (c *DefaultSCCController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("[%s] Starting DefaultSCCController", ControllerName)
+	defer klog.Infof("[%s] Shutting down DefaultSCCController", ControllerName)
+
+	go c.informer.Run(stopCh)
+
+	if !cache.WaitForCacheSync(stopCh, c.informer.HasSynced) {
+		utilruntime.HandleError(fmt.Errorf("[%s] cache for DefaultSCCController did not sync", ControllerName))
+		return
+	}
+
+	go c.runWorker()
+	<-stopCh
+}
+
+func (c *DefaultSCCController) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *DefaultSCCController) processNextWorkItem() bool {
+	key, shutdown := c.queue.Get()
+
+	if shutdown {
+		return false
+	}
+
+	defer c.queue.Done(key)
+
+	request, ok := key.(types.NamespacedName)
+	if !ok {
+		// As the item in the work queue is actually invalid, we call Forget here else
+		// we'd go into a loop of attempting to process a work item that is invalid.
+		c.queue.Forget(key)
+
+		utilruntime.HandleError(fmt.Errorf("[%s] expected types.NamespacedName in workqueue but got %#v", ControllerName, key))
+		return true
+	}
+
+	if err := c.syncer.Sync(request); err != nil {
```

Comment:
do we have this style with a Syncer abstraction elsewhere? Would try to avoid to introduce new one-off patterns for this controller.

Reply:
each controller I see in ckao has a `Sync` function.

Reply:
Sync yes, but a `Syncer` abstraction?

Reply:
"a controller `has` a syncer" is more composable and testable than "a controller `is` a syncer". I tried to keep the concerns separate as they should be but it is not consistent with other controllers at the same time.

I removed it to be consistent for now.